engine/lib/registry/factory.go

```diff
@@ -40,35 +41,45 @@ type WorkerFactory interface {
 }
 
 // WorkerConstructor alias to the function that can construct a WorkerImpl
-type WorkerConstructor func(ctx *dcontext.Context, id libModel.WorkerID, masterID libModel.MasterID, config WorkerConfig) lib.WorkerImpl
+type WorkerConstructor[T lib.WorkerImpl, C any] func(
+	ctx *dcontext.Context, id libModel.WorkerID, masterID libModel.MasterID, config C,
+) T
 
 // SimpleWorkerFactory is a WorkerFactory with built-in JSON codec for WorkerConfig.
-type SimpleWorkerFactory struct {
-	constructor WorkerConstructor
-	configTpi   interface{}
+type SimpleWorkerFactory[T lib.WorkerImpl, C any] struct {
+	constructor WorkerConstructor[T, C]
 }
 
 // NewSimpleWorkerFactory creates a new WorkerFactory.
-func NewSimpleWorkerFactory(constructor WorkerConstructor, configType interface{}) *SimpleWorkerFactory {
-	return &SimpleWorkerFactory{
+func NewSimpleWorkerFactory[T lib.WorkerImpl, Config any](
+	constructor WorkerConstructor[T, Config],
+) *SimpleWorkerFactory[T, Config] {
+	// Config must be a pointer
+	if !isPtr[Config]() {
+		// It's okay to panic here.
+		// The developer who used this function mistakenly should
+		// be able to figure out what happened.
+		log.L().Panic("expect worker's config type to be a pointer")
+	}
+	return &SimpleWorkerFactory[T, Config]{
 		constructor: constructor,
-		configTpi:   configType,
 	}
 }
 
 // NewWorkerImpl implements WorkerFactory.NewWorkerImpl
-func (f *SimpleWorkerFactory) NewWorkerImpl(
+func (f *SimpleWorkerFactory[T, C]) NewWorkerImpl(
 	ctx *dcontext.Context,
 	workerID libModel.WorkerID,
 	masterID libModel.MasterID,
 	config WorkerConfig,
```

Comment:
can we require config is of type C?

Reply:
It seems you cannot implement the `WorkerFactory` interface this way. There has to be a place where you do a `type erasure`.